### PR TITLE
Add dialogue state and dialogue state tracker

### LINKED ISSUE #163: Add dialogue management components
Dialogue management components are needed to create input features for the transformer-based user simulator. For example, one sub-vector of the input account for the first mention of the slot.

### PR DESCRIPTION
**Pull request by NoB0**
This PR add components for dialogue management. Information related to the dialogue state are needed to prepare the input feature for the transformer-based user simulator.

## What's changed?

* Add classes for dialogue state and dialogue state tracker
* Add tests for new classes

Fixes #163

### CODE REVIEW

File: tests/dialogue_management/test_dialogue_state_tracker.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    assert initial_state.belief_state == {}
+    return dst
+
+
+def test_update_state_agent(
+    dialogue_state_tracker: DialogueStateTracker,
+) -> None:
+    """Tests dialogue state update with agent dialogue acts."""
+    dialogue_acts = [
+        DialogueAct(Intent("greets")),
```
Comment (kbalog) [inline tests/dialogue_management/test_dialogue_state_tracker.py:32]: Nit: should it be `greet`?

File: usersimcrs/dialogue_management/dialogue_state.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+"""Representation of the dialogue state.
+
+The dialogue state includes the turn count, a list of dialogue acts per turn for
+both the agent and the user, and the belief state. The belief state is a
+dictionary that maps slot names to slot values.
```
Comment (kbalog) [inline usersimcrs/dialogue_management/dialogue_state.py:5]: Nit: as the purpose of the dictionary is not the "mapping" but to store values provided slots, perhaps phrase it slightly differently as "dictionary that holds for each slot (key) the values provided by the user"?

File: usersimcrs/dialogue_management/dialogue_state.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+"""Representation of the dialogue state.
+
+The dialogue state includes the turn count, a list of dialogue acts per turn for
```
Comment (kbalog) [inline usersimcrs/dialogue_management/dialogue_state.py:3]: (Not a comment for this specific file, but in general) Do we state somewhere how we define turns? I.e., is each utterance by a participant a turn or is an exchange of utterances a turn?

File: usersimcrs/dialogue_management/dialogue_state.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    Attributes:
+        turn_count: Turn count.
+        agent_dacts: List of dialogue acts per turn for the agent.
+        user_dacts: List of dialogue acts per turn for the user.
+        belief_state: Belief state.
+    """
+
+    turn_count: int = 0
+    agent_dacts: List[List[DialogueAct]] = field(default_factory=list)
```
Comment (kbalog) [inline usersimcrs/dialogue_management/dialogue_state.py:27]: Nit: might be worth writing out in full as `[user|agent]_dialogue_acts` for readability

File: usersimcrs/dialogue_management/dialogue_state_tracker.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        """Updates the belief state based on the dialogue acts.
+
+        Args:
+            dialogue_acts: Dialogue acts.
+        """
+        for dialogue_act in dialogue_acts:
+            for annotation in dialogue_act.annotations:
+                self._dialogue_state.belief_state[
+                    annotation.slot
+                ] = annotation.value
```
Comment (kbalog) [inline usersimcrs/dialogue_management/dialogue_state_tracker.py:53]: Do we want to allow only a single value per slot? What if the user wants "action" OR "comedy" genres?

**Review by kbalog — CHANGES_REQUESTED** [review]
Looks very nice! Not approving just yet because of an open question regarding single- vs. multi-valued slots.

File: usersimcrs/dialogue_management/dialogue_state.py (hunk 2 above)
Comment (NoB0) [reply usersimcrs/dialogue_management/dialogue_state.py:5]: Good point, done.

File: usersimcrs/dialogue_management/dialogue_state.py (hunk 4 above)
Comment (NoB0) [reply usersimcrs/dialogue_management/dialogue_state.py:27]: Done.

File: usersimcrs/dialogue_management/dialogue_state.py (hunk 3 above)
Comment (NoB0) [reply usersimcrs/dialogue_management/dialogue_state.py:3]: The docstring is updated to explicitly describe what is a turn.

File: usersimcrs/dialogue_management/dialogue_state_tracker.py (hunk 5 above)
Comment (NoB0) [reply usersimcrs/dialogue_management/dialogue_state_tracker.py:53]: I would say that we should allow for multiple values, this is especially true for slots such as keywords. I changed the type of `belief_state` in `DialogueState` and how it is updated here.

File: tests/dialogue_management/test_dialogue_state_tracker.py (hunk 1 above)
Comment (NoB0) [reply tests/dialogue_management/test_dialogue_state_tracker.py:32]: Done.

File: usersimcrs/dialogue_management/dialogue_state.py [hunk 6]
```diff
@@ -1,8 +1,8 @@
 """Representation of the dialogue state.
 
-The dialogue state includes the turn count, a list of dialogue acts per turn for
-both the agent and the user, and the belief state. The belief state is a
-dictionary that maps slot names to slot values.
+The dialogue state includes the number of utterance, a list of dialogue acts per utterance for both
```
Comment (kbalog) [inline usersimcrs/dialogue_management/dialogue_state.py:3]: Nit: "number of utterances" and check line length

**Review by kbalog — APPROVED** [review]
LGTM

File: usersimcrs/dialogue_management/dialogue_state.py (hunk 6 above)
Comment (NoB0) [reply usersimcrs/dialogue_management/dialogue_state.py:3]: Done.

**Pull request merged**